File: frontend/src/services/finance.ts
import api from './api'

export interface FinanceSource {
  id: string
  name: string
  description?: string | null
  created_at: string
}

export interface FinanceEntry {
  id: string
  user_id: string
  house_id: string
  label: string
  amount: number
  type: 'income' | 'expense'
  subtype: 'recurring' | 'done'
  frequency?: 'monthly' | 'weekly' | 'every_n_months' | null
  frequency_n?: number | null
  date?: string | null
  start_date?: string | null
  end_date?: string | null
  source_id?: string | null
  investment_id?: string | null
  created_at: string
}

export interface FinanceSavings {
  amount: number
  resign_date: string | null
}

export interface RevolutMovement {
  id: string
  user_id: string
  house_id: string
  label: string
  amount: number
  category: string
  date: string
  notes?: string | null
  entity_id?: string | null
  source_id?: string | null
  created_at: string
}

export interface OCRParsed {
  label: string
  amount: number
  type: 'income' | 'expense'
  tx_date?: string | null
  subtype: 'done' | 'recurring'
  entity?: string | null
  source?: string | null
}

export interface FinanceLabel {
  id: string
  name: string
  created_at: string
}

export interface FinanceEntity {
  id: string
  name: string
  category?: string | null
  created_at: string
}

export interface HouseMember {
  user_id: string
  name: string
  monthly_income: number
  monthly_expense: number
  leftover: number
  savings: number
}

export interface HouseSummary {
  members: HouseMember[]
  total_income: number
  total_expense: number
  total_savings: number
  total_leftover: number
}

export const financeService = {
  async listEntries(): Promise<FinanceEntry[]> {
    const res = await api.get('/finance/entries')
    return res.data
  },
  async createEntry(data: {
    label: string
    amount: number
    type: string
    subtype: string
    frequency?: string | null
    frequency_n?: number | null
    date?: string | null
    start_date?: string | null
    end_date?: string | null
    label_id?: string | null
    entity_id?: string | null
    entity_name?: string | null
    source_id?: string | null
    source_name?: string | null
  }): Promise<FinanceEntry> {
    const res = await api.post('/finance/entries', data)
    return res.data
  },
  async deleteEntry(id: string): Promise<void> {
    await api.delete(`/finance/entries/${id}`)
  },
  async updateEntry(id: string, data: { label?: string; amount?: number; start_date?: string | null; end_date?: string | null }): Promise<FinanceEntry> {
    const res = await api.patch(`/finance/entries/${id}`, data)
    return res.data
  },

  async getSavings(): Promise<FinanceSavings> {
    const res = await api.get('/finance/savings')
    return res.data
  },
  async saveSavings(data: FinanceSavings): Promise<FinanceSavings> {
    const res = await api.post('/finance/savings', data)
    return res.data
  },

  async listRevolut(): Promise<RevolutMovement[]> {
    const res = await api.get('/finance/revolut')
    return res.data
  },
  async createRevolut(data: { label: string; amount: number; category: string; date: string; notes?: string; entity_name?: string; entity_id?: string; source_name?: string; source_id?: string }): Promise<RevolutMovement> {
    const res = await api.post('/finance/revolut', data)
    return res.data
  },
  async deleteRevolut(id: string): Promise<void> {
    await api.delete(`/finance/revolut/${id}`)
  },

  async importOcr(file: File): Promise<OCRParsed[]> {
    const form = new FormData()
    form.append('file', file)
    const res = await api.post('/finance/import/ocr', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    return res.data
  },

  async houseSummary(): Promise<HouseSummary> {
    const res = await api.get('/finance/house/summary')
    return res.data
  },

  async listLabels(): Promise<FinanceLabel[]> {
    const res = await api.get('/finance/labels')
    return res.data
  },

  async listEntities(): Promise<FinanceEntity[]> {
    const res = await api.get('/finance/entities')
    return res.data
  },
  async createEntity(data: { name: string; category?: string }): Promise<FinanceEntity> {
    const res = await api.post('/finance/entities', data)
    return res.data
  },
  async deleteEntity(id: string): Promise<void> {
    await api.delete(`/finance/entities/${id}`)
  },

  async listSources(): Promise<FinanceSource[]> {
    const res = await api.get('/finance/sources')
    return res.data
  },
  async createSource(data: { name: string; description?: string }): Promise<FinanceSource> {
    const res = await api.post('/finance/sources', data)
    return res.data
  },
  async deleteSource(id: string): Promise<void> {
    await api.delete(`/finance/sources/${id}`)
  },
}

export default financeService

File: backend/app/models/finance.py
"""
Finance Models
Personal finance tracking: entries (income/expense), savings, Revolut movements.
"""

from sqlalchemy import Column, ForeignKey, String, Date, Numeric, Text, Integer
from sqlalchemy.dialects.postgresql import UUID

from app.models.base import BaseModel


class FinanceLabel(BaseModel):
    __tablename__ = "finance_labels"

    name = Column(String(255), nullable=False, unique=True)


class FinanceEntity(BaseModel):
    __tablename__ = "finance_entities"

    name = Column(String(255), nullable=False, unique=True)
    category = Column(String(80), nullable=True)


class FinanceSource(BaseModel):
    __tablename__ = "finance_sources"

    name = Column(String(255), nullable=False, unique=True)
    description = Column(String(255), nullable=True)


class FinanceEntry(BaseModel):
    __tablename__ = "finance_entries"

    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    house_id = Column(
        UUID(as_uuid=True),
        ForeignKey("houses.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    label = Column(String(255), nullable=False)
    amount = Column(Numeric(12, 2), nullable=False)
    type = Column(String(20), nullable=False)  # income | expense
    subtype = Column(String(20), nullable=False)  # recurring | done
    frequency = Column(String(30), nullable=True)  # monthly | weekly | every_n_months
    frequency_n = Column(Integer, nullable=True)
    date = Column(Date, nullable=True)
    start_date = Column(Date, nullable=True)
    end_date = Column(Date, nullable=True)
    label_id = Column(
        UUID(as_uuid=True),
        ForeignKey("finance_labels.id", ondelete="SET NULL"),
        nullable=True,
    )
    entity_id = Column(
        UUID(as_uuid=True),
        ForeignKey("finance_entities.id", ondelete="SET NULL"),
        nullable=True,
    )
    source_id = Column(
        UUID(as_uuid=True),
        ForeignKey("finance_sources.id", ondelete="SET NULL"),
        nullable=True,
    )
    investment_id = Column(
        UUID(as_uuid=True),
        ForeignKey("investments.id", ondelete="SET NULL"),
        nullable=True,
    )


class FinanceSavings(BaseModel):
    __tablename__ = "finance_savings"

    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    house_id = Column(
        UUID(as_uuid=True),
        ForeignKey("houses.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    amount = Column(Numeric(14, 2), nullable=False, default=0)
    resign_date = Column(Date, nullable=True)


class RevolutMovement(BaseModel):
    __tablename__ = "revolut_movements"

    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    house_id = Column(
        UUID(as_uuid=True),
        ForeignKey("houses.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    label = Column(String(255), nullable=False)
    amount = Column(Numeric(12, 2), nullable=False)
    category = Column(String(80), nullable=False, default="Altro")
    date = Column(Date, nullable=False)
    notes = Column(Text, nullable=True)
    label_id = Column(
        UUID(as_uuid=True),
        ForeignKey("finance_labels.id", ondelete="SET NULL"),
        nullable=True,
    )
    entity_id = Column(
        UUID(as_uuid=True),
        ForeignKey("finance_entities.id", ondelete="SET NULL"),
        nullable=True,
    )
    source_id = Column(
        UUID(as_uuid=True),
        ForeignKey("finance_sources.id", ondelete="SET NULL"),
        nullable=True,
    )
